[FIX] registrations multimap made fully thread safe

romainmoreau opened the pull request:
It used to throw a `ConcurrentModificationException` when there was a high level of concurrency.

Here's a simple demo that reproduces the issue:

```java
import java.util.Collections;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Supplier;
import java.util.stream.IntStream;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import com.google.common.collect.Multimaps;

public class SimpleDemo {
    private static final String KEY = "key";

    public static void main(String[] args) {
        System.out.println("Thread safe demo");
        demo(SimpleDemo::threadSafe);
        System.out.println("Partially thread safe demo");
        demo(SimpleDemo::partiallyThreadSafe);
    }

    private static Multimap<String, Integer> threadSafe() {
        return Multimaps.synchronizedSetMultimap(
                Multimaps.newSetMultimap(new HashMap<>(), () -> Collections.newSetFromMap(new ConcurrentHashMap<>())));
    }

    private static Multimap<String, Integer> partiallyThreadSafe() {
        return Multimaps.synchronizedSetMultimap(HashMultimap.create());
    }

    private static void demo(Supplier<Multimap<String, Integer>> multimapSupplier) {
        var multimap = multimapSupplier.get();
        IntStream.range(0, 128).parallel().forEach(i -> {
            multimap.put(KEY, i);
            multimap.get(KEY).stream().mapToInt(j -> j).sum();
        });
    }
}
```

quantranhong1999 reviewed (approved):
Thanks. Should we have a unit test to reproduce this?

chibenwa commented:
Good catch!

Maybe we could craft a unit test using `ConcurrentTestRunner` against th in vm event bus and try to trigger the bug? (Bonus...)

romainmoreau commented:
Sure I'll try to craft a unit test.

romainmoreau commented:
I added a test that reproduces the issue to the generic concurrent test suite and added this suite to the in VM implementation (not sure why it wasn't already before).
